### ISSUE TITLE
GPU collectives hang

### ISSUE THREAD

**Issue by astooke** [Theano/libgpuarray#387]
Hi,
I've just started to experience a program hanging when calling the NCCL collectives through pygpu.  What I see in nvidia-smi is the utilization of all GPUs pegged at 100% except for one which is at 0%.  But actually the power consumption of all of them is at idle.  And it's not always the same GPU at 0%.  The program just hangs, and it takes a while to kill it.  In some cases it got past the collectives call but then would hang with one process tried to touch the GpuArray variable, either in another function or by calling `np.assary()` on it.  Everything was working recently, so there is probably something I changed that is breaking it, but I'm haven't been able to find it yet....wondering if anyone else has seen anything similar?

I have verified that the processes join the same `clique_id`...and they are all calling `reduce()` with the same shape and data type GpuArray and the same root.  All the NCCL tests are passing.  `pygpu.test()` all passes.  The same code works on a different computer but that's less exciting because it has fewer GPUs.

Driver 375.26 on a DGX-1.

**Comment by abergeron** [comment]
This seems like you don't set the size of your group correctly.  But I don't have your code so I can't know for sure.

**Comment by astooke** [comment]
Ok have just gotten back to this...and I encounter the behavior only when calling `reduce` and only when using more than 2 GPUs.  

Here is a test that runs on my 2 GPU machine and runs when using only 2 GPUs on the 8-GPU machine, but it hangs during the `reduce` call for 3 or more GPUs.  Maybe I am doing something wrong with the `reduce` call?  It also happens when I provide a `dest` and/or `root` arguments in the master.

[gpu_comm_test.txt](https://github.com/Theano/libgpuarray/files/877348/gpu_comm_test.txt)

Like I said, this was working previously, although I didn't track exactly what version.  I have just pulled and rebuilt libgpuarray and it's still happening.

**Comment by astooke** [comment]
OK turns out it was a problem with NCCL, not with pygpu.  

Although the NCCL test for reduce was working fine, I ran into the same problem during their `reduce_scan` test.  So I wiped NCCL, pulled, rebuilt, and now it's all working. :)  

I don't remember changing NCCL before this problem started...there were only a few of the lines in my nccl git pull, mostly in `reduce.cu`... I didn't look inside but maybe they found it.
